docs/manual/src/internals/rendering_foreign_bindings.md

```diff
@@ -2,84 +2,69 @@
-    template which arranges for them to be placed in the correct location.
-
-### Including templates once
-
-We want our type template files to render runtime code, but only once.  For example, we only want to render
-`ObjectRuntime.kt` once, even if there are multiple Object types defined in the UDL file.  To handle this the type
-template defines an `include_once_check()` method, which tests if we've included a file before.  The template code then
-uses that to guard the Askama `{% include %}` statement.  See [`Object.kt` for an
-example](https://github.com/mozilla/uniffi-rs/blob/470740289258e1f06171a976d8e15978f028e391/uniffi_bindgen/src/bindings/kotlin/templates/ObjectTemplate.kt#L2)
+The templates have access to `ci`, a [`ComponentInterface`](https://github.com/mozilla/uniffi-rs/blob/main/uniffi_bindgen/src/interface/mod.rs), which is the Rust representation of all the UniFFI types in your crate.
+
+The task of the templates is to render `ci` into a "module" for the foreign binding.  This mainly consists of rendering support for each [`Type`](https://github.com/mozilla/uniffi-rs/blob/main/uniffi_meta/src/types.rs) described in your crate.
+
+Eg, here's where [Python uses `ci` to iterate over the types](https://github.com/mozilla/uniffi-rs/blob/884f7865f3367c494e9165e21c1255018577db01/uniffi_bindgen/src/bindings/python/templates/Types.py#L3)
+
+The templates create foreign-native types for everything from ffi-native types (int/etc) to functions, dictionaries etc. The implementation of these generated types might call your your Rust implemented FFI, as described below.
+
+Bidings also need to do alot of work to make language identifiers etc work correctly - eg, turn `this_func(this_arg: ThisType)` into `thisFunc(...)`
+
+## Breaking down a Rust function called by Python.
+
+Let's take a quick look at where [Python generates a top-level public function](https://github.com/mozilla/uniffi-rs/blob/884f7865f3367c494e9165e21c1255018577db01/uniffi_bindgen/src/bindings/python/templates/TopLevelFunctionTemplate.py#L37-L40+)
+
+It's making, eg `def this_func(this_arg=0) -> None:` - let's break it down:
+
+`def {{ func.name()|fn_name }}({%- call py::arg_list_decl(func) -%}) -> None:`
+
+The Askama language does funky things with `{ }` blocks and here we are getting string output into the generated code. The first is the `this_func`:
+
+`{{ func.name()|fn_name }}`: [Here is `func.name()`](https://github.com/mozilla/uniffi-rs/blob/884f7865f3367c494e9165e21c1255018577db01/uniffi_bindgen/src/interface/function.rs#L72) - you can see all the other metadata about functions there too.
+The string result from that function goes through an Askama "filter" concept to "pipe" the output to [`fn_name`, this rust function](https://github.com/mozilla/uniffi-rs/blob/884f7865f3367c494e9165e21c1255018577db01/uniffi_bindgen/src/bindings/python/gen_python/mod.rs#L567) - which ends up just handing the fact it might be a Python keyword but otherwise returns the same value.
```

Comment:
```suggestion
Askama uses a "filter" concept: Functions that take the value left of the pipe operator (`|`) to produce a new value.
The "filter" used in the above template is called`fn_name` and is [defined in the Python bindings generator](https://github.com/mozilla/uniffi-rs/blob/884f7865f3367c494e9165e21c1255018577db01/uniffi_bindgen/src/bindings/python/gen_python/mod.rs#L567) - which ends up just handing the fact it might be a Python keyword but otherwise returns the same value.
```